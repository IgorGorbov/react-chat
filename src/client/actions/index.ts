import {
  CLIENT_CONNECT,
  REGISTER_CLIENT,
  FETCH_ALL_USERS,
  FETCH_USER_CHATS,
  SEARCH_CHAT,
  CHANGE_USER_STATUS,
  CHANGE_CHAT,
  ADD_CHAT,
  NEW_CHAT,
  DELETE_CHAT,
  ADD_MESSAGE,
  NEW_MESSAGE,
  CHOOSE_USER,
  SELECT_MESSAGE,
  DELETE_MESSAGES,
  READ_MESSAGES,
  REFRESH_MESSAGES,
} from '../constants/actionTypes';

export const clientConnect = () => ({
  type: CLIENT_CONNECT,
});

export const changeUserStatus = (status: string) => ({
  type: CHANGE_USER_STATUS,
  payload: status,
});

export const changeChat = (chat: { id: number; companion: IUser }) => ({
  type: CHANGE_CHAT,
  payload: {
    id: chat.id,
    companion: chat.companion,
  },
});

export const fetchAllUsers = (users: IUser[]) => ({
  type: FETCH_ALL_USERS,
  payload: users,
});

export const fetchUserChats = (chats: IChat[]) => ({
  type: FETCH_USER_CHATS,
  payload: chats,
});

export const search = (text: string) => ({
  type: SEARCH_CHAT,
  payload: text,
});

export const addChat = (companion: IUser) => ({
  type: ADD_CHAT,
  payload: companion,
});

export const newChat = (currentUser: IUser, companion: IUser) => ({
  type: NEW_CHAT,
  payload: { currentUser, companion },
});

export const deleteChat = (chatId: number) => ({
  type: DELETE_CHAT,
  payload: chatId,
});

export const addMessage = (text: string) => ({
  type: ADD_MESSAGE,
  payload: text,
});

export const newMessage = (chatId: number, message: IMessage) => ({
  type: NEW_MESSAGE,
  payload: {
    chatId,
    message,
  },
});

export const registerClient = (name: string) => ({
  type: REGISTER_CLIENT,
  payload: name,
});

export const chooseUser = (user: IUser) => ({
  type: CHOOSE_USER,
  payload: user,
});

export const selectMessage = (messageId: number) => ({
  type: SELECT_MESSAGE,
  payload: messageId,
});

export const deleteMessages = (
  selectedMessages: number[],
  currentUserID: number,
  currentChatId: number,
) => ({
  type: DELETE_MESSAGES,
  payload: { selectedMessages, currentUserID, currentChatId },
});

export const refreshMessages = (messages: number[], currentChatId: number) => ({
  type: REFRESH_MESSAGES,
  payload: { messages, currentChatId },
});

export const readMessages = (currentChatId: number, currentUserID: number) => ({
  type: READ_MESSAGES,
  payload: { currentUserID, currentChatId },
});
