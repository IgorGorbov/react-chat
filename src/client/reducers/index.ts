import { combineReducers } from 'redux';
import session from './session';
import chats from './chats';
import filters from './filters';
import currentChat from './currentChat';
import users from './users';

export default combineReducers({
  session,
  users,
  chats,
  currentChat,
  filters,
});
