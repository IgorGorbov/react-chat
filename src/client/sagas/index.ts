import { SagaIterator } from 'redux-saga';
import { all, takeEvery, call, select } from 'redux-saga/effects';
import * as types from '../constants/actionTypes';
import {
  getCompanion,
  getChatCurrentId,
  getCurrentUser,
} from '../selectors/index';

const clientConnect = function*(socket: ISocket): SagaIterator {
  yield call(socket.getUsers, 'available', (err: any, users: IUser[]) => {
    if (err) {
      console.error('clientConnect saga => getUsers', err);
      return;
    }
    socket.dispatchAction({ type: types.FETCH_ALL_USERS, payload: users });
  });
};

const registerClient = function*(
  socket: ISocket,
  action: { payload: string },
): SagaIterator {
  yield call(socket.register, action.payload, (err: any, user: IUser) => {
    if (err) {
      console.error('registerClient saga => register', err);
      return;
    }
    socket.dispatchAction({ type: types.CHOOSE_USER, payload: user });
  });

  yield call(socket.getUserChats, (err: any, chats: IChat[]) => {
    if (err) {
      console.error('registerClient saga => getUserChats', err);
      return;
    }
    socket.dispatchAction({ type: types.FETCH_USER_CHATS, payload: chats });
  });

  yield call(socket.getUsers, 'all', (err: any, users: IUser[]) => {
    if (err) {
      console.error('registerClient saga => getUsers', err);
      return;
    }
    socket.dispatchAction({ type: types.FETCH_ALL_USERS, payload: users });
  });
};

const addChat = function*(
  socket: ISocket,
  action: { payload: IUser },
): SagaIterator {
  yield call(socket.addChat, action.payload, (err: any, chat: IChat) => {
    if (err) {
      console.error('addChat saga => addChat', err);
      return;
    }
    socket.dispatchAction({ type: types.NEW_CHAT, payload: chat });
  });
};

const deleteChat = function*(
  socket: ISocket,
  action: { payload: number },
): SagaIterator {
  yield call(socket.deleteChat, action.payload);
};

const addMessage = function*(
  socket: ISocket,
  action: { payload: string },
): SagaIterator {
  const text = action.payload;
  const companion = yield select(getCompanion);
  const chatId = yield select(getChatCurrentId);

  yield call(socket.addMessage, text, chatId, companion);
};

const deleteMessages = function*(
  socket: ISocket,
  action: { payload: number[] },
): SagaIterator {
  const currentChatId = yield select(getChatCurrentId);

  yield call(
    socket.deleteMessages,
    action.payload,
    (err: any, messages: IMessage[]) => {
      if (err) {
        console.error('deleteMessages saga => deleteMessages', err);
        return;
      }
      socket.dispatchAction({
        type: types.REFRESH_MESSAGES,
        payload: { currentChatId, messages },
      });
    },
  );
};

const changeUserStatus = function*(socket: ISocket): SagaIterator {
  const currentUser = yield select(getCurrentUser);
  yield call(socket.changeUserStatus, currentUser);
};

export default function* SocketSaga(socket: ISocket): SagaIterator {
  yield all([
    takeEvery(types.CLIENT_CONNECT, clientConnect, socket),
    takeEvery(types.REGISTER_CLIENT, registerClient, socket),
    takeEvery(types.CHANGE_USER_STATUS, changeUserStatus, socket),
    takeEvery(types.ADD_CHAT, addChat, socket),
    takeEvery(types.DELETE_CHAT, deleteChat, socket),
    takeEvery(types.ADD_MESSAGE, addMessage, socket),
    takeEvery(types.DELETE_MESSAGES, deleteMessages, socket),
  ]);
}
