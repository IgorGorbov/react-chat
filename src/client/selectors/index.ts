import { createSelector } from 'reselect';

const getSearchText = (state: IState) =>
  state.filters.get('search').toLowerCase();

const getChats = (state: IState) => state.chats;

const getCurrentId = (state: IState) => state.session.get('id');

export const getCurrentUser = (state: IState) => state.session.toJS();

export const getChatCurrentId = (state: IState) => state.currentChat.get('id');

export const getCurrentChat = (state: IState) => state.currentChat.toJS();

export const getCompanion = (state: IState) =>
  state.currentChat.get('companion');

export const getSelectedMessages = (state: IState) =>
  state.filters.get('selectedMessages');

export const getSearchEntities: any = createSelector(
  getSearchText,
  getChats,
  getCurrentId,
  (searchText, chats: any, currentID) => {
    const searchByName = (chat: any) => {
      const user = chat.users.find((u: any) => u.id !== currentID);
      return user.name
        .trim()
        .toLowerCase()
        .includes(searchText);
    };
    return chats.filter(searchByName).toJS();
  },
);

export const getChatById: any = createSelector(
  getChats,
  getChatCurrentId,
  (chats, id) => {
    return chats.find((c: any) => c.id === id);
  },
);

const getUsers = (state: any) => state.users;

export const getAvailableUsers: any = createSelector(
  getChats,
  getUsers,
  getCurrentId,
  (chats: any, users, currentUserId) => {
    const filterUsers = Array.from(
      new Set(
        Object.values(chats.toJS()).reduce(
          (acc: any, chat: any) =>
            acc.concat(chat.users.map((user: any) => user.id)),
          [currentUserId],
        ),
      ),
    );

    return users.filter((user: any) => !filterUsers.includes(user.id));
  },
);
