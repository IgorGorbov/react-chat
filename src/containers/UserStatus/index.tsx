import * as React from 'react';
import { Dispatch } from 'redux';
import {
  UserStatusContainer,
  RadioInput,
} from '../../styledComponents'
import { Status } from '../../constants/user'

declare interface IUserStatus {
  changeUserStatus: Dispatch<any>
  userStatus: string,
  handleClickAvatar: () => void
}

class UserStatus extends React.Component<IUserStatus, {}> {

  public handleClickLabel(newStatus: string): void {
    this.props.changeUserStatus(newStatus);
    this.props.handleClickAvatar();
  }

  public render() {
    const { userStatus } = this.props;
    return (
      <UserStatusContainer>
        <RadioInput
          className="online"
          type="radio"
          id="r1"
          name="status"
          defaultChecked={userStatus === Status.Online}
        />
        <label
          onClick={ () => this.handleClickLabel(Status.Online) }
          htmlFor="r1"
        >
          Online
        </label>

        <RadioInput
          className="away"
          type="radio"
          id="r2"
          name="status"
          defaultChecked={userStatus === Status.Away}
        />
        <label
          onClick={ () => this.handleClickLabel(Status.Away) }
          htmlFor="r2"
        >
          Away
        </label>

        <RadioInput
          className="busy"
          type="radio"
          id="r3"
          name="status"
          defaultChecked={userStatus === Status.Busy}
        />
        <label
          onClick={ () => this.handleClickLabel(Status.Busy) }
          htmlFor="r3"
        >
          Busy
        </label>

        <RadioInput
          className="offline"
          type="radio"
          id="r4"
          name="status"
          defaultChecked={userStatus === Status.Offline}
        />
        <label
          onClick={ () => this.handleClickLabel(Status.Offline) }
          htmlFor="r4"
        >
          Offline
        </label>
      </UserStatusContainer>
    );
  }
}

export default UserStatus;
