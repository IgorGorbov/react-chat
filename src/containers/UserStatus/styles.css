.CheckUserStatus {
    display: flex;
    flex-direction: column;
    position: absolute;
    top: 115px;
    left: 10px;
    width: 150px;
    height: 160px;
    z-index: 20;
    background-color: #485F78;
    box-shadow: 0 19px 38px rgba(0,0,0,0.30), 0 15px 12px rgba(0,0,0,0.22);
}

.CheckUserStatus:before {
    content: '';
    position: absolute;
    display: block;
    left: 50px;
    top: -10px;
    width: 0;
    height: 0;
    border-left: 10px solid transparent;
    border-right: 10px solid transparent;
    border-bottom: 10px solid #485F78;
}

.radio {
    position: absolute;
    z-index: -1;
    opacity: 0;
    margin: 10px 0 0 7px;
}
.radio + label {
    position: relative;
    padding: 10px 0 10px 55px;
    cursor: pointer;
    color: white;
    font-family: Titillium Web,sans-serif;
    font-size: 15px;
    transition: background-color 0.3s ease;
}
.radio + label:hover {
    background-color: #32465a;
}
.radio + label:before {
    content: '';
    position: absolute;
    top: 13px;
    left: 15px;
    width: 14px;
    height: 14px;
    border: 1px solid #485F78;
    border-radius: 50%;
    background: green;
}
.radio + label:after {
    content: '';
    position: absolute;
    top: 16px;
    left: 17px;
    width: 9px;
    height: 9px;
    border-radius: 50%;
    opacity: 0;
    transition: .3s ease;
}
.radio:checked + label:after {
    opacity: 1;
}
.radio:focus + label:before {
    box-shadow: 0 0 0 2px rgba(72, 95, 120,.9), 0 0 0 4px rgba(0,128,0, 1);
}


.radio.busy + label:before {
    background: red;
}

.radio.busy:focus + label:before {
    box-shadow: 0 0 0 2px rgba(72, 95, 120,.9), 0 0 0 4px rgba(255, 0, 0,.8);
}

.radio.away + label:before {
    background: yellow;
}

.radio.away:focus + label:before {
    box-shadow: 0 0 0 2px rgba(72, 95, 120,.9), 0 0 0 4px rgba(255,255,0,.9);
}

.radio.offline + label:before {
    background: #E7EAEA;
}

.radio.offline:focus + label:before {
    box-shadow: 0 0 0 2px rgba(72, 95, 120,.9), 0 0 0 4px rgba(231,234,234,.9);
}
