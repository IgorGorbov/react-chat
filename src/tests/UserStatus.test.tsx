import * as React from 'react';
import toJSON from 'enzyme-to-json';
import { mount } from 'enzyme';

import UserStatus from '../client/components/UserStatus';

import { UserStatusContainer, RadioInput } from '../client/styledComponents';

describe('testing UserStatus', () => {
  interface IUserStatus {
    changeUserStatus: (status: string) => { type: string; payload: string };
    userStatus: string;
    handleClickAvatar: () => void;
  }
  const FakeChangeUserStatus = jest.fn();
  const FakeHandleClickAvatar = jest.fn();

  const initState: IUserStatus = {
    changeUserStatus: FakeChangeUserStatus,
    userStatus: 'Online',
    handleClickAvatar: FakeHandleClickAvatar,
  };

  let wrapper: any;

  beforeAll(() => {
    wrapper = mount(<UserStatus {...initState} />);
  });

  it('should contain UserStatusContainer', () => {
    expect(wrapper.find(UserStatusContainer)).toHaveLength(1);
  });

  it('should contain RadioInput', () => {
    expect(wrapper.find(RadioInput)).toHaveLength(4);
  });

  it('should contain label', () => {
    const labels = wrapper.find('label');
    expect(labels).toHaveLength(4);

    labels.first().simulate('click');
    expect(FakeChangeUserStatus).toHaveBeenCalledTimes(1);
    expect(FakeHandleClickAvatar).toHaveBeenCalledTimes(1);
  });

  it('snapshot UserStatus', () => {
    const tree = toJSON(wrapper);
    expect(tree).toMatchSnapshot();
  });
});
