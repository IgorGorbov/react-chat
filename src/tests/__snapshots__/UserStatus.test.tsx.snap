// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing UserStatus snapshot UserStatus 1`] = `
<UserStatus
  changeUserStatus={
    [MockFunction] {
      "calls": Array [
        Array [
          "Online",
        ],
      ],
    }
  }
  handleClickAvatar={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
    }
  }
  userStatus="Online"
>
  <styled.div>
    <div
      className="sc-ckVGcZ iLgDVT"
    >
      <styled.input
        className="online"
        defaultChecked={true}
        id="r1"
        name="status"
        type="radio"
      >
        <input
          className="online sc-jKJlTe feDwuL"
          defaultChecked={true}
          id="r1"
          name="status"
          type="radio"
        />
      </styled.input>
      <label
        htmlFor="r1"
        onClick={[Function]}
      >
        Online
      </label>
      <styled.input
        className="away"
        defaultChecked={false}
        id="r2"
        name="status"
        type="radio"
      >
        <input
          className="away sc-jKJlTe feDwuL"
          defaultChecked={false}
          id="r2"
          name="status"
          type="radio"
        />
      </styled.input>
      <label
        htmlFor="r2"
        onClick={[Function]}
      >
        Away
      </label>
      <styled.input
        className="busy"
        defaultChecked={false}
        id="r3"
        name="status"
        type="radio"
      >
        <input
          className="busy sc-jKJlTe feDwuL"
          defaultChecked={false}
          id="r3"
          name="status"
          type="radio"
        />
      </styled.input>
      <label
        htmlFor="r3"
        onClick={[Function]}
      >
        Busy
      </label>
      <styled.input
        className="offline"
        defaultChecked={false}
        id="r4"
        name="status"
        type="radio"
      >
        <input
          className="offline sc-jKJlTe feDwuL"
          defaultChecked={false}
          id="r4"
          name="status"
          type="radio"
        />
      </styled.input>
      <label
        htmlFor="r4"
        onClick={[Function]}
      >
        Offline
      </label>
    </div>
  </styled.div>
</UserStatus>
`;
